libbpf-tools/profile.bpf.c

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2021 Wenbo Zhang
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "profile.h"
+#include "maps.bpf.h"
+
+#define MAX_ENTRIES		10240
+
+const volatile bool kernel_stacks_only = false;
+const volatile bool user_stacks_only = false;
+const volatile bool include_idle = false;
+const volatile pid_t targ_tgid = -1;
+const volatile pid_t targ_pid = -1;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__uint(key_size, sizeof(u32));
+} stackmap SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, struct key_t);
+	__type(value, sizeof(u64));
+	__uint(max_entries, MAX_ENTRIES);
+} counts SEC(".maps");
+
+SEC("perf_event")
+int do_perf_event(struct bpf_perf_event_data *ctx) {
+	u64 id = bpf_get_current_pid_tgid();
+	u32 tgid = id >> 32;
+	u32 pid = id;
+	u64 *valp;
+	static const u64 zero;
+
+	if (include_idle == false && pid == 0)
+		return 0;
+
+	if (targ_tgid != -1 && targ_tgid != tgid)
+		return 0;
+	if (targ_pid != -1 && targ_pid != pid)
+		return 0;
+
+	// create map key
+	struct key_t key = {.pid = tgid};
```

Comment:
Declare variables first.

This `.pid = tgid` is very counterintuitive. I would suggest you use `pid` and `tid` instead.